tensorflow/core/kernels/fixed_length_record_reader_op.cc

```diff
@@ -29,73 +32,137 @@ class FixedLengthRecordReader : public ReaderBase {
-    // File: "0123456"
-    // Reader setting: `record_bytes=3`, `hop_bytes=2`, `footer_bytes=0`,
-    //     `header_bytes=0`
-    // Without this checking condition, the forth time the reader will at
-    // this position: "012345|6" and the reading operation will result in
-    // an error.
-    if (input_buffer_->Tell() >= file_pos_limit_ ||
-        input_buffer_->Tell() + record_bytes_ > file_pos_limit_) {
+    // In case hop_bytes_ is in between 0 and record_bytes_,
+    // we will need to hold a cache so that later on we could prefix the cache
+    // with the remaining data to read:
+    // For example, assume record_bytes is 2 and hop_bytes is 1,
+    // For file "H123",
+    // We will process the data in the following way:
+    // 1. Read hope_bytes of 1 ("1") and save it in hop cache.
+    // 2. Read record_bytes - hop_bytes = 1 ("2")
+    // 3. Prefix data in 2. with hop cache in one, we have "12",
+    //    this is used for the record.
+    // 4. Shift hop_bytes of 1 from the record, and we put "2" to hop cache.
+    // 5. Continue step 2.
+    // In order to acheive the above, in the following only
+    // 'record_bytes_ - lookahead_cache_.size()' needs to be read. The lookahead_cache_
+    // is then prefixed to piece together the whole record.
+
+    int bytes_to_read = record_bytes_;
+    if (0 < hop_bytes_ && hop_bytes_ < record_bytes_) {
+        bytes_to_read = hop_bytes_;
+    }
+
+    // In case of non-seekable files like compressed file, attempt to read
+    // record, and append record to cached footer, then pop head
+    // of cached footer as the record
+    Status s = buffered_input_stream_->ReadNBytes(bytes_to_read, value);
+    if (!s.ok() && !errors::IsOutOfRange(s)) {
+      return s;
+    }
+
+    // EOF reached
+    if (value->size() == 0) {
       *at_end = true;
       return Status::OK();
     }
-    const int64 pos_before_read = input_buffer_->Tell();
-    TF_RETURN_IF_ERROR(input_buffer_->ReadNBytes(record_bytes_, value));
+
+    // We have the data which is pieced together by
+    // lookahead_cache_ + value(read)
+    lookahead_cache_.append(*value, 0, bytes_to_read);
+
+    // Now the true data will be:
+    // value: the first record_bytes_
+    // hop cache (part of lookahead_cache_): the next chunk starts
+    //    with hop_bytes and ends with record_bytes
+    //    (total = record_bytes_ - hop_bytes_)
+    // footer cache (part of lookahead_cache_): the final chunk of
+    //    size footer_bytes
+    *value = lookahead_cache_.substr(0, record_bytes_);
+    lookahead_cache_ = lookahead_cache_.substr(bytes_to_read);
+
     *key = strings::StrCat(current_work(), ":", record_number_);
     *produced = true;
     ++record_number_;
 
-    if (hop_bytes_ > 0) {
-      input_buffer_->Seek(pos_before_read + hop_bytes_).IgnoreError();
+    if (hop_bytes_ > record_bytes_) {
+      buffered_input_stream_->SkipNBytes(hop_bytes_ - record_bytes_);
```

Comment:
Thanks @saxenasaurabh. Yes the skip should happen at the beginning of the `lookahead_cache_` and `buffered_input_stream_` should read and append buffer to `lookahead_cache_`. I have updated the implementation and added the test cases for it.